src/OrbitGl/Batcher.h

```diff
@@ -135,6 +135,8 @@ class Batcher {
               std::unique_ptr<PickingUserData> user_data = nullptr);
   void AddBox(const Box& box, const Color& color, std::shared_ptr<Pickable> pickable);
 
+  void AddBoxBorder(const Box& box, const Color& color, std::shared_ptr<Pickable> pickable);
```

Comment:
I'm probably already fighting a lost cause, but I'm really not a fan of adding all of these methods to add more complex primitives to the Batcher. Initially, the Batcher supported the core Primitives used in rendering: Lines, Triangles, Boxes. It's clear how they are rendered, and it's defined that for a same layer, we'll first draw boxes, then lines, then primitives.

IMO the methods that add more complex elements really should not live in the Batcher, because it's not the responsibility of the Batcher to assemble primitives to complex elements. But this isn't the first method to do this - this already happened with AddCircle, AddShadedBox, AddShadedTrapezium... but all of those use the public Batcher API, so they really could all be free functions, or live in another class.

*If* you add new methods, then they should support the same overloads as the primitive operations - as user data, they accept `std::unique_ptr<PickingUserData>`, `std::shared_ptr<Pickable>`, or `nullptr`. This has also not been done for the new methods that were added. I assume you only need the weird `NonPickable` workaround because you only defined the `std::shared_ptr<Pickable>` overload of this, or was there another reason? This looks like a hack, I don't like to introduce this. Our picking is (overly) complicated as it already is :-)

Reply:
I'll move the `AddBoxBorder` from `Batcher`. 

My first version of `AddBoxBorder` was passing `nullptr` instead of `std::shared_ptr<Pickable>`. This lead to deselection of the timer when the user click right on border. `NonPickable` allows to circumvent this issue via doing nothing.

Reply:
I don't know what should be the long-term plan, but I would rather think that I prefer to have all this functions together. It's easier to find an issue if these complex functions live in the same place and also easy to move them all together in the future. So, I would prefer to have this in batcher for now.

And @akopich, why not having a shared pointer between the border and the box itself so if you select the border or the center, it highlights the entire box? For example:
- If (ShouldHaveBorder) {
   AddBoxWithBorder(..., std::shared_ptr<Pickable> pickable);
} 
else {
   AddBox()
}

Or this doesn't work? I don't remember if we are already sharing picking_id between elements. Could you confirm me that this is possible @reichlfl ??

Reply:
`PickingUserData` can be passed to the methods of `Batcher` only via shared ptr.